Editor/Prefab Handling/PrefabFolderStripper.cs

```diff
@@ -0,0 +1,146 @@
+            if (StripSettings.StripFoldersFromPrefabsInBuild)
+                StripFoldersFromDependentPrefabs();
+        }
+
+        public void OnPostprocessBuild(BuildReport report)
+        {
+            if (StripSettings.StripFoldersFromPrefabsInBuild)
+                RevertChanges();
+        }
+
+        private static void HandlePrefabsOnPlayMode(PlayModeStateChange state)
+        {
+            if ( ! StripSettings.StripFoldersFromPrefabsInPlayMode || StripSettings.PlayMode == StrippingMode.DoNothing)
+                return;
+
+            if (state == PlayModeStateChange.ExitingEditMode)
+            {
+                // Stripping folders from all prefabs in the project instead of only the ones referenced in the scenes
+                // because a prefab can be hot-swapped in Play Mode.
+                StripFoldersFromAllPrefabs();
+            }
+            else if (state == PlayModeStateChange.EnteredEditMode)
+            {
+                RevertChanges();
+            }
+        }
+
+        private static void StripFoldersFromDependentPrefabs()
+        {
+            var scenePaths = EditorBuildSettings.scenes.Select(scene => scene.path).ToArray();
+            var dependentAssetsPaths = AssetDatabase.GetDependencies(scenePaths, true);
+
+            var prefabsWithLabel = dependentAssetsPaths.Where(path =>
+                    AssetDatabase.GetLabels(GetAssetForLabel(path)).Contains(LabelHandler.FolderPrefabLabel))
+                .ToArray();
+
+            _changedPrefabs = new (string, string)[prefabsWithLabel.Length];
+
+            for (int i = 0; i < prefabsWithLabel.Length; i++)
+            {
+                string path = prefabsWithLabel[i];
+                _changedPrefabs[i] = (path, File.ReadAllText(path));
+                StripFoldersFromPrefab(path, StripSettings.Build);
+            }
+        }
+
+        private static
+#if UNITY_2020_1_OR_NEWER
+            GUID
+#else
+            Object
+#endif
+            GetAssetForLabel(string path)
+        {
+#if UNITY_2020_1_OR_NEWER
+            return AssetDatabase.GUIDFromAssetPath(path);
+#else
+            return AssetDatabase.LoadAssetAtPath<GameObject>(path);
+#endif
+        }
+
+        private static void StripFoldersFromAllPrefabs()
+        {
+            var prefabGUIDs = AssetDatabase.FindAssets($"l: {LabelHandler.FolderPrefabLabel}");
+            _changedPrefabs = new (string, string)[prefabGUIDs.Length];
+
+            for (int i = 0; i < prefabGUIDs.Length; i++)
+            {
+                string guid = prefabGUIDs[i];
+                string path = AssetDatabase.GUIDToAssetPath(guid);
+
+                _changedPrefabs[i] = (path, File.ReadAllText(path));
+                StripFoldersFromPrefab(path, StripSettings.PlayMode);
+            }
+        }
+
+        private static void StripFoldersFromPrefab(string prefabPath, StrippingMode strippingMode)
+        {
+            using (var temp = new EditPrefabContentsScope(prefabPath))
```

Comment:
![image](https://user-images.githubusercontent.com/9505910/108790131-9484fe80-7541-11eb-889a-34522b8afaa2.png)

Having a folder as the root of a prefab causes the flattening procedure to destroy all children. Maybe it should leave it for when instantiated, then flatten itself onto the hierarchy as expected?

(2019.4, 2020.1.4f1, 2021.1.0b7)

Reply:
If the prefab will flatten upon instantiation, it will be against the purpose of the plugin which is to improve the game performance. Flattening will slow down the game instead of speeding it up. I suggest the following alternative:
1. If the folder is a root of the prefab, a warning will be output, saying that the hierarchy will not flatten for this prefab.
2. The Folder component will be stripped from the folder game object, so just an empty game object will be left as a prefab root.

What do you think?

Reply:
That sounds good to me. Let's do it that way.

Reply:
Added the following warning:

> Hierarchy will not flatten for {prefabPath} because its root is a folder. It's advised to make the root an empty game object.

Let me know if the phrasing is fine.